### PR TITLE
Getting started refactor

### PR DESCRIPTION
**Pull request by tcare**
To address feedback about some confusion when following the guide (especially to do with service connections) I have refactored the document structure.
- Optional tasks / extra info has been moved down to either 'Further Exploration' (directly for getting started) or 'Next Steps' (specifically for integrating ML code.)
- Grammar fixes and simplified wording in parts
- Removed manual word wrapping to let autoformat do its thing
- Added a table of contents

### CODE REVIEW

File: docs/getting_started.md [hunk 1]
```diff
@@ -1,219 +1,176 @@
...
+* The first stage of the pipeline, **Model CI**, performs linting, unit testing, code coverage, building and publishes an **ML Training Pipeline** in an **ML Workspace**.
 
 * The second stage of the pipeline, **Train model**, triggers the run of the ML Training Pipeline. The training pipeline will train, evaluate, and register a new model. The actual computation is performed in an [Azure Machine Learning Compute cluster](https://docs.microsoft.com/en-us/azure/machine-learning/service/how-to-set-up-training-targets#amlcompute). In Azure DevOps, this stage runs an agentless job that waits for the completion of the Azure ML job, allowing the pipeline to wait for training completion for hours or even days without using agent resources.
 
-**Note:** If the model evaluation determines that the new model does not perform better than the previous one then the new model will not be registered and the pipeline will be cancelled.
+**Note:** If the model evaluation determines that the new model does not perform better than the previous one, the new model will not be registered and the pipeline will be cancelled.
```
Comment (dtzar) [inline docs/getting_started.md]: Please add the step which will fail, the error message, and possibly the link in the evaulate code too, so we don't get _another_ issue filed on this topic.

File: docs/getting_started.md [hunk 2]
```diff
@@ -1,219 +1,176 @@
...
 
-![trained model](./images/trained-model.png)
+![Trained model](./images/trained-model.png)
 
 To disable the automatic trigger of the training pipeline, change the `auto-trigger-training` variable as listed in the `.pipelines\diabetes_regression-ci.yml` pipeline to `false`.  This can also be overridden at runtime execution of the pipeline.
 
 To skip model training and registration, and deploy a model successfully registered by a previous build (for testing changes to the score file or inference configuration), add the variable `MODEL_BUILD_ID` when the pipeline is queued, and set the value to the id of the previous build.
 
-### Deploy the Model to Azure Kubernetes Service
+## Futher Exploration
```
Comment (dtzar) [inline docs/getting_started.md]: Would be nice to have a short sentence with what is here and plus potentially link or mention of the sections here.

File: docs/getting_started.md [hunk 3]
```diff
@@ -1,219 +1,176 @@
...
 
-The final stage is to deploy the model to the production environment running on
-[Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service).
+### (Optional) Deploy the Model to Azure Kubernetes Service
```
Comment (dtzar) [inline docs/getting_started.md]: Please give some explanation of ACI is already deployed because it is lightweight approach which allows an easy/quick smoke test. They should only really need one deployment option, etc....

File: docs/getting_started.md [hunk 4]
```diff
@@ -1,219 +1,176 @@
...
 
-**Note:** Creating a Kubernetes cluster on AKS is out of scope of this
-tutorial, but you can find set up information
-[here](https://docs.microsoft.com/en-us/azure/aks/kubernetes-walkthrough-portal#create-an-aks-cluster).
+**Note:** Creating a Kubernetes cluster on AKS is out of scope of this tutorial, but you can find set up information [here](https://docs.microsoft.com/en-us/azure/aks/kubernetes-walkthrough-portal#create-an-aks-cluster).
 
 **Note:** If your target deployment environment is a K8s cluster and you want to implement Canary and/or A/B testing deployment strategies check out this [tutorial](./canary_ab_deployment.md).
 
+The final stage is to deploy the model to the production environment running on
```
Comment (dtzar) [inline docs/getting_started.md]: final stage in the sample pipeline to illustrate all potential deployment choices..  Possibly this is addressed by the section above about the deployment options.

File: docs/getting_started.md [hunk 5]
```diff
@@ -282,3 +233,35 @@ Make sure your webapp has the credentials to pull the image from the Azure Conta
...
   * **Distributed tracing**: The smoke test client code sets an HTTP `traceparent` header (per the [W3C Trace Context proposed specification](https://www.w3.org/TR/trace-context-1)), and the `score.py` code logs this header. The query above shows how to surface this value. You can adapt this to your tracing framework.
   * **Monitoring**: You can use [Azure Monitor for containers](https://docs.microsoft.com/en-us/azure/azure-monitor/insights/container-insights-overview) to monitor the Azure ML scoring containers' performance, just as for any other container.
+
+### Clean up the example resources
+
+(Optional) To remove the resources created for this project you can use the [/environment_setup/iac-remove-environment-pipeline.yml](../environment_setup/iac-remove-environment-pipeline.yml) definition or you can just delete the resource group in the [Azure Portal](https://portal.azure.com).
+
+## Next Steps: Integrating your project
+
+* The provided pipeline definition YAML files are a sample starting point, which you should tailor to your processes and environment.
```
Comment (dtzar) [inline docs/getting_started.md]: Remove this line.  Should be covered by the bootstrap.

File: docs/getting_started.md [hunk 6]
```diff
@@ -282,3 +233,35 @@ Make sure your webapp has the credentials to pull the image from the Azure Conta
...
+* Follow the [bootstrap instructions](../bootstrap/README.md) to create a starting point for your project use case.
+* Use the [Convert ML experimental code to production code](https://docs.microsoft.com/azure/machine-learning/tutorial-convert-ml-experiment-to-production#use-your-own-model-with-mlopspython-code-template) tutorial which explains how to bring your machine learning code on top of this template.
```
Comment (dtzar) [inline docs/getting_started.md]: Remove, covered in bootstrap.

File: docs/getting_started.md [hunk 7]
```diff
@@ -282,3 +233,35 @@ Make sure your webapp has the credentials to pull the image from the Azure Conta
...
+* You may want to use [Azure DevOps self-hosted agents](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/agents?view=azure-devops&tabs=browser#install) to speed up your ML pipeline execution. The Docker container image for the ML pipeline is sizable, and having it cached on the agent between runs can trim several minutes from your runs.
+* You can install additional Conda or pip packages by modifying the YAML environment configurations under the `diabetes_regression` directory. Make sure to use fixed version numbers for all packages to ensure reproducibility, and use the same versions across environments.
```
Comment (dtzar) [inline docs/getting_started.md]: Remove, should be covered by the self-hosted environments doc linked to in the bootstrap.

File: docs/getting_started.md [hunk 8]
```diff
@@ -282,3 +233,35 @@ Make sure your webapp has the credentials to pull the image from the Azure Conta
...
+* You should edit the pipeline definition to remove unused stages. For example, if you are deploying to ACI and AKS only, you should delete the unused `Deploy_Webapp` stage.
```
Comment (dtzar) [inline docs/getting_started.md]: Move this guidance to the section you need to add above where we talk about the three options.

File: docs/getting_started.md [hunk 9]
```diff
@@ -282,3 +233,35 @@ Make sure your webapp has the credentials to pull the image from the Azure Conta
...
+* You may wish to enable [manual approvals](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/approvals) before the deployment stages.
```
Comment (dtzar) [inline docs/getting_started.md]: Move this into the AKS section.  You can only do this with AKS deployments today.

File: docs/getting_started.md [hunk 10]
```diff
@@ -282,3 +233,35 @@ Make sure your webapp has the credentials to pull the image from the Azure Conta
...
+
+### Using an existing dataset
+
+**Note:** The training ML pipeline uses a [sample diabetes dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html) as training data. To use your own data, you need to [create a Dataset](https://docs.microsoft.com/azure/machine-learning/how-to-create-register-datasets) in your workspace and specify its name in a DATASET_NAME variable in the ***devopsforai-aml-vg*** variable group. You will also need to modify the test cases in the **ml_service/util/smoke_test_scoring_service.py** script to match the schema of the training features in your dataset.
```
Comment (dtzar) [inline docs/getting_started.md]: Move this section to the bootstrap guide.

File: docs/getting_started.md [hunk 11]
```diff
@@ -1,238 +1,204 @@
...
+  - [Additional Variables and Configuration](#additional-variables-and-configuration)
+    - [More variable options](#more-variable-options)
+    - [Local configuration](#local-configuration)
 
-* Fork this repository if there is a desire to contribute back to the repository else
-* Use this [code template](https://github.com/microsoft/MLOpsPython/generate) which copies the entire code base to your own GitHub location with the git commit history restarted. This can be used for learning and following the guide.
+## Setting up Azure DevOps
 
-This repository contains a template and demonstrates how to apply it to a sample ML project ***diabetes_regression*** that creates a linear regression model to predict the diabetes.
+You'll use Azure DevOps for running the multi-stage pipeline with build (CI), model training, and scoring service release (CD) stages. If you don't already have an Azure DevOps organization, create one by following the instructions at [Quickstart: Create an organization or project collection](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/create-organization?view=azure-devops).
```
Comment (vtbassmatt) [inline docs/getting_started.md]: I would leave out the "(CI)" and "(CD)" indicators. CI isn't exactly equivalent to build, and CD is a particular form of release strategy. The difference isn't all that meaningful to someone just trying to get started with MLOps.

File: docs/getting_started.md [hunk 12]
```diff
@@ -1,238 +1,204 @@
...
+The variable group should contain the following required variables:
 
-**Note:** Creating the ARM service connection scope requires 'Owner' or 'User Access Administrator' permissions on the subscription.
-You must also have sufficient permissions to register an application with
-your Azure AD tenant, or receive the ID and secret of a service principal
-from your Azure AD Administrator. That principal must have 'Contributor'
-permissions on the subscription.
+| Variable Name            | Suggested Value           | Short description                                                                                                            |
+| ------------------------ | ------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
+| BASE_NAME                | [unique base name]        | Naming prefix                                                                                                                |
```
Comment (vtbassmatt) [inline docs/getting_started.md]: Should this suggest an example name?

File: docs/getting_started.md [hunk 13]
```diff
@@ -1,238 +1,204 @@
...
-| BASE_NAME                | [unique base name]       |
-| LOCATION                 | centralus                |
-| RESOURCE_GROUP           | mlops-RG                 |
-| WORKSPACE_NAME           | mlops-AML-WS             |
-| AZURE_RM_SVC_CONNECTION  | azure-resource-connection|
-| WORKSPACE_SVC_CONNECTION | aml-workspace-connection |
-| ACI_DEPLOYMENT_NAME      | diabetes-aci             |
-
-**Note:**
+If you intend to contribute back to this repository, fork the project. Otherwise use the [code template](https://github.com/microsoft/MLOpsPython/generate), which copies the entire code base to your own GitHub location with the git commit history restarted. You can use the resulting repository for this guide and for your own experimentation.
```
Comment (vtbassmatt) [inline docs/getting_started.md]: We found that giving simpler, "one way to do it" instructions increased completion/success rate. Since forks are so cheap on GitHub, most (all?) Azure DevOps docs unconditionally recommend forking.

File: docs/getting_started.md [hunk 14]
```diff
@@ -1,238 +1,204 @@
...
-Azure resources. When the solution is used in a shared subscription, there can
-be naming collisions with resources that require unique names like azure blob
-storage and registry DNS naming. Make sure to give a unique value to the
-BASE_NAME variable (e.g. MyUniqueML), so that the created resources will have
-unique names (e.g. MyUniqueMLamlcr, MyUniqueML-AML-KV, etc.). The length of
-the BASE_NAME value should not exceed 10 characters and it should contain numbers and letters only.
+You'll need to create a *variable group* in Azure DevOps to store values that are reused across multiple pipelines or pipeline stages. Either store the values directly in [Azure DevOps](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=designer#create-a-variable-group) or connect to an Azure Key Vault in your subscription. Check out the [Add & use variable groups](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=yaml#use-a-variable-group) documentation to learn more about how to create a variable group and link it to your pipeline.
```
Comment (vtbassmatt) [inline docs/getting_started.md]: Might be nice to have a second sentence which says "The example pipeline expects a variable group in order to set some key values" (or something like that) - basically, why am I doing this?

Comment (tcare) [reply docs/getting_started.md]: Agreed. We also want to reference the section at the bottom with more info on the complete list of variables and how they can be toggled

File: docs/getting_started.md (hunk 13 above)
Comment (tcare) [reply docs/getting_started.md]: I agree we should only give one way to do it. IMO, people that want to contribute back will figure out they need to fork, so I'll remove that bit for now.

Code template vs. forking is out of scope of this PR, but an interesting discussion. One advantage of forking would be to easily get updates, especially since you don't get a choice to base the template on a specific release version. With #219 merged we'll have even more separation of the infrastructure code from the project code, which allows for easier downstream updates without conflicts. @dtzar, any thoughts?

Comment (dtzar) [reply docs/getting_started.md]: @vtbassmatt I generally agree.  However this IS a `code template` and is intended for someone to use for this primary purpose and are equally as "cheap" and quick to do.  Also, if they want to use this project for their own code, a fork will send them down a bad path where they'll have to start over with a code template vs easy transition from getting started.

Comment (vtbassmatt) [reply docs/getting_started.md]: Good point! I was mostly referring to "give them one way to do it". (GitHub has template repos now - not sure if you've seen that feature.)

File: docs/getting_started.md (hunk 12 above)
Comment (tcare) [reply docs/getting_started.md]: We had one at some point (I think 'diabetes') but wanted to avoid having diabetes-specific instructions if possible. I'll think about an alternative.

File: docs/getting_started.md (hunk 13 above)
Comment (dtzar) [reply docs/getting_started.md]: This is a GitHub template repository - is there something I'm missing?

Comment (tcare) [reply docs/getting_started.md]: Maybe using 'repository template' is better since that's the official GitHub terminology

Comment (dtzar) [reply docs/getting_started.md]: Also I wonder if it would help if we state it like this instead "which effectively forks the entire repository plus resets the git commit history."

File: docs/getting_started.md (hunk 12 above)
Comment (tcare) [reply docs/getting_started.md]: I ended up going with the intention of the name (your project name) rather than a specific example. I also added the restrictions as a quick reference.
